### PR TITLE
IGNITE-19730 Sql. ExecutionServiceImplTest different tests failed

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

**Comment by zstan** [comment]
@korlov42 tests are not flaky any more, as for me code become a little bit clear for understanding. can u make a review plz ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 1]
```diff
@@ -824,47 +824,62 @@ private void enlist(SourceAwareIgniteRel rel) {
...
-
-                        var finalStepFut = cancelResult.whenComplete((r, e) -> {
-                            queryManagerMap.remove(ctx.queryId());
-
-                            try {
-                                ctx.cancel().cancel();
-                            } catch (Exception ignored) {
-                                // NO-OP
-                            }
+            .thenCompose(tmp -> {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:835]: indentation

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 2]
```diff
@@ -888,14 +903,6 @@ private CompletableFuture<Void> awaitFragmentInitialisationAndClose() {
...
                 cancelFuts.add(
                         CompletableFuture.allOf(entry.getValue().toArray(new CompletableFuture[0]))
-                                .handle((none2, t) -> {
-                                    // Some fragments may be completed exceptionally, and that is fine.
-                                    // Here we need to make sure that all query initialisation requests
-                                    // have been processed before sending a cancellation request. That's
-                                    // why we just ignore any exception.
-
-                                    return null;
-                                })
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:898]: why did you remove this?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java [hunk 3]
```diff
@@ -690,7 +689,7 @@ public void closeSession() {
 
         assertThrowsWithCause(
                 () -> await(ars0.fetchNextPage()),
-                ExecutionCancelledException.class
+                CursorClosedException.class
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java:692]: and what about already exist :
https://github.com/apache/ignite-3/blob/cae90410a32ba3946d09f6a135683f9c3e231f38/modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java#L636
https://github.com/apache/ignite-3/blob/cae90410a32ba3946d09f6a135683f9c3e231f38/modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlSynchronousApiTest.java#L330

[the remark this answers is not in the document]
Comment (korlov42) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java:692]: what is wrong with these two assertion?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java (hunk 2 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:898]: ))) as i can observe through history initially this section process exception handling and after just degenerate into empty block. I agree To fill the issue for this case if you ok.

Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:898]: have you read comment in this empty block?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItCommonApiTest.java [hunk 4]
```diff
@@ -107,7 +105,7 @@ public void testSessionExpiration() throws Exception {
             while (rs1.hasNext()) {
                 rs1.next();
             }
-        }, ExecutionCancelledException.class);
+        }, CursorClosedException.class);
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItCommonApiTest.java]: got it, thanks for pointing, bloody bloody code )

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java (hunk 2 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:898]: of course, but still have no clue, i found that no one test will touch this place with not null **t**

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 5]
```diff
@@ -826,45 +826,69 @@ private void enlist(SourceAwareIgniteRel rel) {
...
+                if (!cancel) {
+                    CompletableFuture<Void> closeFut = node.closeAsync();
+
+                    return start.thenCompose(v -> closeFut);
+                }
+
+                node.onError(new ExecutionCancelledException());
+            }
+
+            return start;
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:883]: is there any reason to move this method up? If no, let's just revert this change

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:898]: it's not news that we don't have comprehensive test coverage.

In the very beginning, the comment states "Some fragments may be completed exceptionally". Therefore, we need to find out whether this invariant is conserved. There are only 5 usages `remoteFragmentInitCompletion`, so I believe it's not a rocket science to take a look at every usage. As far as I can see, `DistributedQueryManager#onNodeLeft` completes init future with exception. 

Feel free to add such test to ExecutionServiceImplTest. If it will pass without mentioned `handle` block, I'm ok to remove it

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 6]
```diff
@@ -826,45 +826,69 @@ private void enlist(SourceAwareIgniteRel rel) {
...
-                        return cancelResult.thenCombine(finalStepFut, (none1, none2) -> null);
-                    });
+                    ctx.cancel().cancel();
+
+                    cancelFut.complete(null);
+                });
+
+                return cancelResult.thenCombine(finalStepFut, (none1, none2) -> null);
+            })
+                    .thenRun(() -> localFragments.forEach(f -> f.context().cancel()));
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:855]: any reason to reformat this part?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 7]
```diff
@@ -838,33 +838,34 @@ private CompletableFuture<Void> close(boolean cancel) {
...
-                            try {
-                                ctx.cancel().cancel();
-                            } catch (Exception ignored) {
-                                // NO-OP
+                                LOG.warn("Exception raised during close", ex);
                             }
 
+                            queryManagerMap.remove(ctx.queryId());
+
+                            ctx.cancel().cancel();
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:849]: why did you removed try-catch block around cancel?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 8]
```diff
@@ -838,33 +838,34 @@ private CompletableFuture<Void> close(boolean cancel) {
...
                         var finalStepFut = cancelResult.whenComplete((r, e) -> {
-                            queryManagerMap.remove(ctx.queryId());
+                            if (e != null) {
+                                Throwable ex = unwrapCause(e);
 
-                            try {
-                                ctx.cancel().cancel();
-                            } catch (Exception ignored) {
-                                // NO-OP
+                                LOG.warn("Exception raised during close", ex);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:844]: Log message should be rephrased.

According to our [guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Logging+Guidelines#JavaLoggingGuidelines-WARN):
```
WARN is what the team will need to review at the start of the next business day, at a high priority. WARN is always actionable - the user must understand how to react
```

From the message I don't understand what I should to do. TBH, I don't even understand what was going wrong. We closed a door and exception had been raised? A window?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 9]
```diff
@@ -838,33 +838,38 @@ private CompletableFuture<Void> close(boolean cancel) {
                                 : closeLocalFragments();
 
                         var finalStepFut = cancelResult.whenComplete((r, e) -> {
+                            if (e != null) {
+                                Throwable ex = unwrapCause(e);
+
+                                LOG.warn(format("Fragment closing processed with errors, root fragmentId={}", rootFragmentId), ex);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:844]: I've got a few questions regarding this line of code:
1) why don't delegate formatting to logger?
2) how did you come up with rootFragmentId param?
3) have you read logging guidelines?

Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java]: ok, i change it

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 10]
```diff
@@ -838,33 +838,38 @@ private CompletableFuture<Void> close(boolean cancel) {
                                 : closeLocalFragments();
 
                         var finalStepFut = cancelResult.whenComplete((r, e) -> {
+                            if (e != null) {
+                                Throwable ex = unwrapCause(e);
+
+                                LOG.warn("Fragment closing processed with errors, root [queryId={}]", ex, ctx.queryId());
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:844]: not sure why do we need "root" at the end

Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java]: it`s all about "self review" )

**Review by korlov42 — APPROVED** [review]
